master/internal/api_experiment.go

```diff
@@ -0,0 +1,168 @@
+package internal
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+var errExperimentNotFound = status.Error(codes.NotFound, "experiment not found")
+
+func canActivate(dbExp *model.Experiment) error {
+	if dbExp.State != model.PausedState {
+		return status.Error(codes.InvalidArgument, "experiment is not in a paused state")
+	}
+	return nil
+}
+
+func canPause(dbExp *model.Experiment) error {
+	if dbExp.State != model.ActiveState { // TODO does this make sense?
+		return status.Error(codes.InvalidArgument, "experiment is not in a active state")
+	}
+	return nil
+}
+
+func canCancel(dbExp *model.Experiment) error {
+	if _, exist := model.RunningStates[dbExp.State]; !exist {
+		return status.Error(codes.InvalidArgument, "experiment is not in a running state")
+	}
+	return nil
+}
+
+func canArchive(dbExp *model.Experiment) error {
+	if dbExp.Archived {
+		return status.Error(codes.InvalidArgument, "experiment is already archived")
+	}
+	return nil
+}
+
+func canUnArchive(dbExp *model.Experiment) error {
+	if !dbExp.Archived {
+		return status.Error(codes.InvalidArgument, "experiment is not archived")
+	}
+	return nil
+}
+
+type validityCheck func(a *model.Experiment) error
+
+func isValidOperation(dbConn *db.PgDB, expID int, checker validityCheck) (*model.Experiment, error) {
+
+	dbExp, err := dbConn.ExperimentByID(expID)
+
+	switch {
+	case err == db.ErrNotFound:
+		return nil, errUserNotFound
+	case err != nil:
+		return nil, errors.Wrapf(err, "loading experiment %v", expID)
+	}
+
+	err = checker(dbExp)
+	if err != nil {
+		return nil, err
+	}
+
+	return dbExp, nil
+}
+
+func (a *apiServer) ActivateExperiment(
+	ctx context.Context, req *apiv1.SetExperimentAttributeRequest) (*apiv1.SetExperimentAttributeResponse, error) {
+
+	expID := (int)(req.Id) // platform dependent int
```

Comment:
would we need to guard against overflow?!

Reply:
no